File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
n = int(input())
d = []

for i in range(n):
    d.append(list(map(int, input().split())))

dp = [[0] * (i + 1) for i in range(n)]
dp[0][0] = d[0][0]

for i in range(1, n):
    for j in range(i + 1):
        if j == 0:
            dp[i][j] = dp[i - 1][j] + d[i][j]
        elif j == i:
            dp[i][j] = dp[i - 1][j - 1] + d[i][j]
        else:
            dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + d[i][j]

print(max(dp[n - 1]))

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim();

function solution(input) {
  let result = 0;

  const parseInput = input.split(`\n`);
  const N = Number(parseInput[0]);
  const triangle = parseInput.slice(1).map(a => a.split(' ').map(Number));

  for (let i = N - 1; i >= 0; i--) {
    for (let j = 0; j < triangle[i].length - 1; j++) {
      triangle[i - 1][j] = Math.max(
        triangle[i - 1][j] + triangle[i][j],
        triangle[i - 1][j] + triangle[i][j + 1]
      );
    }
  }

  return Number(triangle[0]);
}

console.log(solution(input));
